common/djangoapps/contentserver/middleware.py

```diff
@@ -146,8 +124,75 @@ def process_request(self, request):
             response['Content-Type'] = content.content_type
             response['Last-Modified'] = last_modified_at_str
 
+            # Potentially set content caching headers.
+            self.set_caching_headers(content, response)
+
             return response
 
+    def set_caching_headers(self, content, response):
+        """
+        Sets caching headers based on whether or not the asset is locked.
+        """
+
+        is_locked = getattr(content, "locked", False)
+
+        # We want to signal to the end user's browser, and to any intermediate proxies/caches,
+        # whether or not this asset is cacheable.  If we have a TTL configured, we inform the
+        # caller, for unlocked assets, how long they are allowed to cache it.  Since locked
+        # assets should be restricted to enrolled students, we simply send headers that
+        # indicate there should be no caching whatsoever.
+        cache_ttl = CourseAssetCacheTtlConfig.get_cache_ttl()
+        if cache_ttl > 0 and not is_locked:
+            expire_dt = datetime.datetime.utcnow() + datetime.timedelta(seconds=cache_ttl)
+            response['Expires'] = expire_dt.strftime("%a, %d %b %Y %H:%M:%S GMT")
+            response['Cache-Control'] = "public, max-age={ttl}, s-maxage={ttl}".format(ttl=cache_ttl)
+        elif is_locked:
+            response['Cache-Control'] = "private, no-cache, no-store"
+
+        remove_headers_from_response(response, "Vary")
+
+    def is_user_authorized(self, request, content, location):
+        """
+        Determines whethe or not the user for this request is authorized to view the given asset.
+        """
+
+        is_locked = getattr(content, "locked", False)
+        if is_locked:
+            if not hasattr(request, "user") or not request.user.is_authenticated():
+                return False
+            if not request.user.is_staff:
+                deprecated = getattr(location, 'deprecated', False)
+                if deprecated and not CourseEnrollment.is_enrolled_by_partial(request.user, location.course_key):
+                    return False
+                if not deprecated and not CourseEnrollment.is_enrolled(request.user, location.course_key):
+                    return False
```

Comment:
In general, please try to structure your returns so that it doesn't nest as much. For example:

```python
is_locked = getattr(content, "locked", False)

is not is_locked:
    return True

if not hasattr(request, "user") or not request.user.is_authenticated():
    return False
```